Labels with dictionary

ktsaou opened the pull request:
This PR hides all the complexity about labels, into rrdlabels.c
It uses dictionaries to simplify labels management.

underhood commented on aclk/schema-wrappers/node_info.cc:
```diff
@@ -4,9 +4,15 @@
 
 #include "schema_wrapper_utils.h"
 
+static int label_add_to_map(const char *name, const char *value, LABEL_SOURCE ls, void *data) {
```
since we have same thing in `aclk/schema-wrappers/chart_stream.cc` should we move this function to `schema_wrapper_utils` file? This way we can have don't need to have duplicates

thiagoftsm replied:
@ktsaou, please, take a look in this suggestion.

ktsaou replied:
done

underhood commented:
needs either rebase or merge due to conflicts with master

ktsaou commented:
Guys this needs really a lot of testing.
I have verified the basics, but I have changed really a lot of places!

The good thing, is that the code is now a lot simpler to use and understand and also all the labels complexity should be in rrdlabels.c.

There are a couple of warnings I added and a couple of TODOs. All of them should be settled before merging. But I need your help.

underhood commented on database/rrdlabels.c:
```diff
@@ -50,7 +54,8 @@
 
-#define LABELS_MAX_LENGTH 400 // 400 in bytes, up to 200 UTF-8 characters
+#define LABELS_MAX_NAME_LENGTH 200
+#define LABELS_MAX_VALUE_LENGTH 400 // 400 in bytes, up to 200 UTF-8 characters
```
From this comment and later bellow we have a comment `// UTF-8 encoded 2-byte character` 

it seems we assume UTF-8 being 2 bytes per character. This is not true UTF-8 char can be 1-4 Bytes long

ktsaou replied:
fixed it

ktsaou commented on database/rrdlabels.c:
```diff
@@ -360,60 +365,84 @@ __attribute__((constructor)) void initialize_labels_keys_char_map(void) {
+            else {
+                // UTF-8 characters are not allowed.
+                // Assume it is an underscore
+                // and skip the second byte
```
fixed it

underhood commented on aclk/schema-wrappers/schema_wrapper_utils.h:
```diff
@@ -3,6 +3,8 @@
 #ifndef SCHEMA_WRAPPER_UTILS_H
 #define SCHEMA_WRAPPER_UTILS_H
 
+#include "node_info.h"
```
scope can be limited here to `rrd.h` AFAIK we only need RRDLABEL_SRC

ktsaou replied:
fixed it

ktsaou commented on database/rrd.h:
```diff
@@ -3,6 +3,10 @@
 #define NETDATA_RRD_H 1
 
+#ifdef __cplusplus
+extern "C" {
```
yes, it is included by c++ header files

underhood commented:
just for the record on the ubuntu 18.04 the preprocesor output is:
```
struct disconnect_cmd *parse_disconnect_cmd(const char *data, size_t len) {
    DisconnectReq req;
    struct disconnect_cmd *res;

    if (!req.ParseFromArray(data, len))
        return
# 53 "aclk/schema-wrappers/connection.cc" 3 4
              __null
# 53 "aclk/schema-wrappers/connection.cc"
                  ;

    res = (struct disconnect_cmd *)calloc(1, sizeof(struct disconnect_cmd));

    if (!res)
        return
# 58 "aclk/schema-wrappers/connection.cc" 3 4
              __null
# 58 "aclk/schema-wrappers/connection.cc"
                  ;

    res->reconnect_after_s = req.reconnect_after_seconds();
    res->permaban = req.permaban();
    res->error_code = req.error_code();
    if (req.
# 63 "aclk/schema-wrappers/connection.cc" 3 4
           (json_tokener_errors[])
# 63 "aclk/schema-wrappers/connection.cc"
                              .c_str()) {
        res->error_description = strdup(req.
# 64 "aclk/schema-wrappers/connection.cc" 3 4
                                           (json_tokener_errors[])
# 64 "aclk/schema-wrappers/connection.cc"
                                                              .c_str());
        if (!res->error_description) {
            free(res);
            return
# 67 "aclk/schema-wrappers/connection.cc" 3 4
                  __null
# 67 "aclk/schema-wrappers/connection.cc"
                      ;
        }
    }

    return res;
}
```

so `res->error_description = strdup(req.error_description().c_str());` turns out to be:
```
        res->error_description = strdup(req.
# 64 "aclk/schema-wrappers/connection.cc" 3 4
                                           (json_tokener_errors[])
# 64 "aclk/schema-wrappers/connection.cc"
                                                              .c_str());
```

but only on some systems (older ones which are failing CI)

underhood commented:
https://github.com/json-c/json-c/issues/135

this is what caused the problem

thiagoftsm commented:
Hello @underhood,

I was digging the issue with @ktsaou and we identified that ancient versions of `json-c` are creating the issue, because they have a macro with the same name of our `method`. Some tests I ran shows that version `json-c-json-c-0.14-20200419` already had the issue fixed, so the problems are related to old versions like `0.12` that is installed with Debian buster.

thiagoftsm commented:
> [json-c/json-c#135](https://github.com/json-c/json-c/issues/135)
> 
> this is what caused the problem

Exactly! :)
Inside this file we have the following declaration:

```c
#define error_description(error)  (json_tokener_errors[error])
```

ktsaou commented:
I added this `libnetdata/json/json.h`:

```
#if ENABLE_JSONC
#include <json-c/json.h>
// fix an older json-c bug
// https://github.com/json-c/json-c/issues/135
#ifdef error_description
#undef error_description
#endif // error_description
#endif // ENABLE_JSONC
```

It fixes the problem

thiagoftsm commented on web/api/formatters/json_wrapper.c:
```diff
@@ -207,8 +210,7 @@ void rrdr_json_wrapper_begin(RRDR *r, BUFFER *wb, uint32_t format, RRDR_OPTIONS
             while (pattern && (label_key = simple_pattern_iterate(&pattern))) {
-                uint32_t key_hash = simple_hash(label_key);
-                struct label *current_label;
+                const char *label_value;
```
Hello @ktsaou,

I am testing your PR on different scenarios and I observed that during compilation time I am receiving this warning:

```sh
web/api/formatters/json_wrapper.c: In function 'rrdr_json_wrapper_begin':
web/api/formatters/json_wrapper.c:213:29: warning: unused variable 'label_value' [-Wunused-variable]
  213 |                 const char *label_value;
      |      
```
Do you still intend to use this variable? Or can we remove?

ktsaou replied:
thanks! removed it!

thiagoftsm commented on database/rrdlabels.c:
```diff
@@ -3,201 +3,1138 @@
+
+    size_t sources = 0;
+    if(lb->label_source & RRDLABEL_SRC_AUTO)
+        buffer_sprintf(wb, "%sauto", sources++?",":"");
```
Hello @ktsaou,

I decided to run `coverity-scan` on your PR due the complexity it has, and I received this report:

```c
*** CID 353213:  Control flow issues  (DEADCODE)
/database/rrdlabels.c: 829 in rrdlabels_log_label_to_buffer_callback()
823         RRDLABEL *lb = (RRDLABEL *)value;
824     
825         buffer_sprintf(wb, "Label: %s: \"%s\" (", name, lb->value);
826     
827         size_t sources = 0;
828         if(lb->label_source & RRDLABEL_SRC_AUTO)
>>>     CID 353213:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression "","" inside this statement: "buffer_sprintf(wb, "%sauto"...".
829             buffer_sprintf(wb, "%sauto", sources++?",":"");
830     
831         if(lb->label_source & RRDLABEL_SRC_CONFIG)
832             buffer_sprintf(wb, "%snetdata.conf", sources++?",":"");
833     
834         if(lb->label_source & RRDLABEL_SRC_K8S)

** CID 353212:  Concurrent data access violations  (MISSING_LOCK)
/mqtt_websockets/src/mqtt_ng.c: 389 in buffer_purge()

````

ktsaou replied:
Well, coverity is wrong. Use `netdatacli` to reload the labels to see it working.

thiagoftsm replied:
All right, I am marking as resolved, because the proposed test worked.

thiagoftsm commented on database/rrdlabels.c:
```diff
@@ -3,201 +3,1138 @@
+        t->errors++;
+    }
+    else if(strcmp(value, t->expected_value) != 0) {
+        fprintf(stderr, "values don't match, found \"%s\", expected \"%s\"", value?value:"(null)", t->expected_value?t->expected_value:"(null)");
```
I also received this warning:

```c
*** CID 353214:  Control flow issues  (DEADCODE)
/database/rrdlabels.c: 965 in rrdlabels_unittest_add_a_pair_callback()
959         }
960         else if(value == NULL || t->expected_value == NULL) {
961             fprintf(stderr, "value is wrong, found \"%s\", expected \"%s\"", value?value:"(null)", t->expected_value?t->expected_value:"(null)");
962             t->errors++;
963         }
964         else if(strcmp(value, t->expected_value) != 0) {
>>>     CID 353214:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression ""(null)"" inside this statement: "fprintf(stderr, "values don...".
965             fprintf(stderr, "values don't match, found \"%s\", expected \"%s\"", value?value:"(null)", t->expected_value?t->expected_value:"(null)");
966             t->errors++;
967         }
968     
969         return 1;
970     }

** CID 353213:  Control flow issues  (DEADCODE)
/database/rrdlabels.c: 829 in rrdlabels_log_label_to_buffer_callback()
```

ktsaou replied:
This is a unit test. Yes, if there is no bug, this part will never run.
But the whole point is to catch the error if it happens.

thiagoftsm commented:
I finished a sequence of tests with this PR and I observed the following results:

| Test | Result |
|-------|-----------|
|Compilation on FreeBSD | OK |
|Compilation on Slackware Current | OK |
|Compilation on Arch Linux | OK |
|Access to web API using [our script](https://github.com/netdata/netdata/blob/master/tests/urls/request.sh.in) to do 96417 different requests | OK |
|UTF8 Label | OK |
|Alarm without `foreach` | OK |
|Alarm with`foreach` | OK |
|Alarm template without `foreach` | OK |
|Alarm template with`foreach` | OK |
|Charts on cloud | OK |
|Parent <-> Child streaming with this PR | OK |
|Parent <-> Child (master branch) streaming | OK |
|Cgroup.plugin (Cgroup V1) | OK|
|Cgroup.plugin (Cgroup V2) | OK|
|Registry | OK |
|Exporting (JSON) | OK |
|Exporting Prometheus (2.35) | OK |
|Exporting OpenTSDB HTTP (InfluxDB 1.7.6) | Cannot send data |
|Exporting OpenTSDB Telnet (InfluxDB 1.7.6)  | Cannot send data |

As everybody can see, except for OpenTSDB/InfluxDB the results were fine. @vlvkobal I am not sure if my influx has problems or we really have problems with OpenTSDB, please, can you also test this?

thiagoftsm commented on aclk/schema-wrappers/schema_wrapper_utils.cc:
```diff
@@ -13,3 +13,10 @@ void set_timeval_from_google_timestamp(const google::protobuf::Timestamp &ts, st
 }
+
+int label_add_to_map_callback(const char *name, const char *value, RRDLABEL_SRC ls, void *data) {
+    (void)ls;
```
A minor suggestion, we have defined a macro [UNUSED](https://github.com/netdata/netdata/blob/e61633a9e99e97cf325514b96f08ccafdb640674/libnetdata/libnetdata.h#L301) that inserts a code like this, I suggest to use it.

thiagoftsm replied:
I do not remember who brought it for us, but since it was merged, we have used it hundred of times:

```sh
bash-5.1$ grep -r -w "UNUSED" *  | wc -l
205
```

My suggestion was in the direction to keep an unique pattern, but  I understand your issues. Let us keep your PR as it is, and we can think about other PR to return for the explicit method.  :)

vlvkobal commented:
> I am not sure if my influx has problems or we really have problems with OpenTSDB, please, can you also test this?

You probably tested before [`42601d7` (#13070)](https://github.com/netdata/netdata/pull/13070/commits/42601d75a304721c4556924a8d4cb95376a11bc5) was applied.

vlvkobal commented:
While [workaround for cmocka and strdupz()](https://github.com/netdata/netdata/pull/13070/commits/7bebee04801db1865c748a7896d5fa54bb7104a5) helps the exporting unit tests pass, the more general and proper fix is in https://github.com/ktsaou/netdata/pull/6.

ktsaou commented:
> While [workaround for cmocka and strdupz()](https://github.com/netdata/netdata/pull/13070/commits/7bebee04801db1865c748a7896d5fa54bb7104a5) helps the exporting unit tests pass, the more general and proper fix is in [ktsaou#6](https://github.com/ktsaou/netdata/pull/6).

merged it

ktsaou commented:
So, are we good to merge?

ilyam8 commented:
I tested the following combination in my K8s cluster:

| Parent      | Children    |  Result|
| ----------- | ----------- |------- |
| this branch      | this branch       | OK     |
| this branch   | master OR v1.34.1        | OK     |
| master OR v1.34.1   | this branch        | OK     |

thiagoftsm commented:
> > I am not sure if my influx has problems or we really have problems with OpenTSDB, please, can you also test this?
> 
> You probably tested before [`42601d7` (#13070)](https://github.com/netdata/netdata/pull/13070/commits/42601d75a304721c4556924a8d4cb95376a11bc5) was applied.

It is possible, because I compiled the code early and I began to do the tests without update the branch, I will take a look  again with latest commits.

ktsaou commented:
@stelfrag I put the fix you applied in this PR.

ktsaou commented:
After all our tests, I guess this is good to go. Right?

thiagoftsm reviewed (approved):
After last commits, I ran again all tests that I had described [here](https://github.com/netdata/netdata/pull/13070#issuecomment-1149317196) and I had success with all.

I suggest we expect results from @ilyam8 and @stelfrag , because they are testing on different environments.

stelfrag commented:
Changing `space` and `( )` to underscore makes the host labels in `/api/v1/info` to report

```
"host_labels": {
   "_os_version":"22.04_LTS__Jammy_Jellyfish_"
}
```

instead of 
```
"host_labels": {
   "_os_version": "22.04 LTS (Jammy Jellyfish)",
}
```

In both cases the `os_version` in the /api/v1/info remains the same `"os_version": "22.04 LTS (Jammy Jellyfish)"`
